LayoutTests/fast/attachment/resources/attachment-test-utils.js

```diff
@@ -0,0 +1,51 @@
+function takeScreenshotWhenAttachmentsSettled(message, iteration = 0)
+{
+    console.assert(document.documentElement.classList.contains("reftest-wait"));
+
+    const attachments = [...document.getElementsByTagName("attachment")];
+    console.assert(attachments.length);
+    if (!attachments.length)
+        return document.documentElement.classList.remove("reftest-wait");
+
+    let states = attachments.map(e => [e, []]);
+
+    const timeoutId = setTimeout(function() {
+        for (const [attachment, events] of states) {
+            if (events.at(-1) != "load")
+                attachment.insertAdjacentText("afterend", "<-(" + message + ")-events=[" + events.join() + "]");
+        }
+        requestAnimationFrame(() => {
+            requestAnimationFrame(() => {
+                document.documentElement.classList.remove("reftest-wait");
+            });
+        });
+    }, 5000);
+
+    for (let state of states) {
+        for (const eventType of ["beforeload", "loadingerror", "loadeddata", "error"])
+            state[0].addEventListener(eventType, function() { state[1].push(eventType); });
+
+        state[0].addEventListener("load", function() {
+            state[1].push("load");
+            if (states.every(state => state[1].at(-1) == "load")) {
+                clearTimeout(timeoutId);
+                // Ensure the image has had time to be rendered before taking the screenshot.
+                requestAnimationFrame(() => {
+                    requestAnimationFrame(() => {
+                        document.documentElement.classList.remove("reftest-wait");
+                    });
+                });
+            }
+        });
+    }
```

Comment:
I would write this whole function as:

```js
function takeScreenshotWhenAttachmentsSettled() {
    console.assert(document.documentElement.classList.contains("reftest-wait"));

    const attachments = [...document.getElementsByTagName("attachment")];
    console.assert(attachments.length);
    if (!attachments.length)
        return document.documentElement.classList.remove("reftest-wait");

    const states = new Map(attachments.map(a => [a, []]));

    const onFailure = () => {
        for (const [attachment, events] of states)
            attachment.insertAdjacentText("afterend", "<- (" + message + ") - events = [" + events.join() + "]");
        document.documentElement.classList.remove("reftest-wait");
    };

    const timeoutId = setTimeout(onFailure, 5000);

    const onSuccess = () => {
        clearTimeout(timeoutId);
        document.documentElement.classList.remove("reftest-wait");
    };

    const promises = attachments.map((attachment) => new Promise((resolve, reject) => {
        for (const event of ["beforeload", "loadingerror", "loadeddata", "error"])
            attachment.addEventListener(event, () => { states.get(attachment).push(event); });
        attachment.addEventListener("loadingerror", reject, { once: true });
        attachment.addEventListener("load", () => {
            requestAnimationFrame(() => requestAnimationFrame(resolve));
        }, { once: true });
    }));

    Promise.all(promises).then(onSuccess, onFailure);
}
```

(You don't need the double rAF on failure, and generally this looks cleaner using promises)
(I also added early exit if we reach loadingerror, instead of waiting for the timeout, but feel free to remove that if you prefer).

Reply:
Thank you. The only change I've added, is to also push "load": In case there are multiple attachments, we'll know those have have finished their load.